### PR TITLE
[mono] Don't call Assembly.CodeBase directly in RuntimeAssembly.GetName

### LINKED ISSUE #54835: Mono's RuntimeAssembly.GetName calls CodeBase
CodeBase is marked as `RequiresAssemblyFiles`, which means that it can throw an exception if the application is published as a single-file application.

https://github.com/dotnet/runtime/blob/bde0322a9e11fba40a7a65c6d38200d6514efc72/src/mono/System.Private.CoreLib/src/System/Reflection/RuntimeAssembly.cs#L253-L256

We should decide what to do here for Mono. In CoreCLR, we don't call CodeBase directly, but instead we call the underlying `GetCodeBase()` method, which returns `null` when running in a single-file application:

https://github.com/dotnet/runtime/blob/bde0322a9e11fba40a7a65c6d38200d6514efc72/src/coreclr/System.Private.CoreLib/src/System/Reflection/RuntimeAssembly.cs#L102-L117

Maybe we should do the same?

cc @akoeplinger @steveisok

### PR DESCRIPTION
**Pull request by lambdageek**
It's marked as not available in single file apps.
Call the underlying get_code_base icall.

Fixes https://github.com/dotnet/runtime/issues/54835

### CODE REVIEW

File: src/mono/System.Private.CoreLib/src/System/Reflection/RuntimeAssembly.cs [hunk 1]
```diff
@@ -252,7 +252,7 @@ private static void AddPublicNestedTypes(Type type, List<Type> types, List<Excep
 
         public override AssemblyName GetName(bool copiedName)
         {
-            return AssemblyName.Create(_mono_assembly, CodeBase);
+            return AssemblyName.Create(_mono_assembly, get_code_base (this, false));
```
Comment (marek-safar) [inline src/mono/System.Private.CoreLib/src/System/Reflection/RuntimeAssembly.cs:255]: Is this an actual fix? What does it return for single-file like targets?

Comment (lambdageek) [reply src/mono/System.Private.CoreLib/src/System/Reflection/RuntimeAssembly.cs:255]: In Mono we have two names associated with an assembly.  It's "filename" which corresponds to `Assembly.Location` which is NULL for single-file targets like WebAssembly.  And "name" which is derived from the name that the assembly was called in the bundle which is returned for `Assembly.CodeBase`.

So yea this is the actual fix - the issue is that the CodeBase public getter is tagged as not available in single-file.  But `get_code_base` is not.  And it returns a reasonable value.  So we use it directly.

Comment (marek-safar) [reply src/mono/System.Private.CoreLib/src/System/Reflection/RuntimeAssembly.cs:255]: Yeah and I don't think that's correct. The value is exposed via https://github.com/dotnet/runtime/blob/main/src/libraries/System.Private.CoreLib/src/System/Reflection/AssemblyName.cs#L63 which says that the values should be empty string (null value effectively)

Comment (lambdageek) [reply src/mono/System.Private.CoreLib/src/System/Reflection/RuntimeAssembly.cs:255]: Thanks.  Switched to using `MonoImage:filename` which will be null for bundles

**Review by marek-safar — APPROVED** [review]

**Comment by lambdageek** [comment]
Ok, so now wasm is failing a codebase test in `src/libraries/System.Reflection/tests/AssemblyTests.cs` because we now return null for the codebase here.

```
        [Fact]
        public void CodeBase()
        {
            Assert.NotEmpty(Helpers.ExecutingAssembly.CodeBase);
        }
```

How does this test pass on CoreCLR single file? or is it skipped somehow?  @eerhardt

**Comment by eerhardt** [comment]
> How does this test pass on CoreCLR single file?

I don't believe the test is executed on CoreCLR single file. Honestly, I don't think any of our library tests are executed under CoreCLR single file.

cc @agocke @tlakollo @vitek-karas

**Comment by tlakollo** [comment]
> I don't believe the test is executed on CoreCLR single file. Honestly, I don't think any of our library tests are executed under CoreCLR single file.

I believe that is controlled by the following switch -> https://github.com/dotnet/runtime/blob/main/src/libraries/tests.proj#L329
We only tests single file in System.Collections.Tests.csproj and System.IO.IsolatedStorage.Tests.csproj but could be enabled in other projects

**Comment by marek-safar** [comment]
> How does this test pass on CoreCLR single file? or is it skipped somehow?

I believe there is a single-file like PlatformDetection property.

> We only tests single file in System.Collections.Tests.csproj and System.IO.IsolatedStorage.Tests.csproj but could be enabled in other projects

I think System.Reflection would be a very useful addition to the list.

**Review by eerhardt — APPROVED** [review]

**Pull request merged**